File: lod.r
library(dplyr)

lod_calc <- function(n, cv, beta, k) {
  if (cv == 0) {
    return(-log(beta) / (n * k))
  }
  d <- 1 / cv ^ 2
  ((d / beta ^ (1 / (n * d))) - d) / k
}

calc <- function(cv, beta, k, threshold, arr) {
  lod <- data.frame(n = seq(from = 1, to = 1000, by = 1))
  lod <- mutate(lod, LOD = sapply(lod$n, lod_calc, cv = cv, beta = beta, k = k))
  lod <- mutate(lod, LOD2 = sapply(lod$n, lod_calc, cv = 0, beta = beta, k = k))

  table <- list()
  below_threshold <- lod[lod$LOD < threshold, ]
  if (length(below_threshold[[1]]) == 0) {
    table["Minimum n"] <- "> 1000"
  } else {
    table["Minimum n"] <- below_threshold[1, ]$n
    table["LOD"] <- below_threshold[1, ]$LOD
  }
  ret <- list()
  ret$Graph <- lod[lod$n <= 50, ]
  ret$Graph1 <- lod[lod$n <= 50, ]
  ret$Table <- table
  ret$Table2 <- lod[lod$n <= 10, ]
  ret$Text <- "test"
  ret
}

File: lod.py
import pandas as pd
import numpy as np

def lod_calc(n, cv, beta, k):
    if cv == 0:
        return -np.log(beta) / (n * k)
    d = 1 / np.pow(cv,2)
    return ((d / np.pow(beta, 1 / (n * d))) - d) / k

def calc(num, cv, beta, k):
    lod = pd.DataFrame(np.linspace(1, 1000, num=1000), columns=['n'])
    for i in range(num):
        lod[f'Line {i}'] = lod.n.apply(lod_calc, args=[cv[i], beta[i], k[i]])

    return {'Graph': lod[lod['n'] <= 50], 'Table': lod[lod['n'] <= 10]}